standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -1201,22 +1199,86 @@ private TableValidWriteIds getValidWriteIdsForTable(Statement stmt, String fullT
     }
   }
 
+  private List<Long> getValidTxnIds(AllocateTableWriteIdsRequest rqst, Statement stmt, ResultSet rs)
+          throws MetaException, SQLException {
+    List<Long> txnIds = new ArrayList<>();
+    String dbName = rqst.getDbName().toLowerCase();
+    String tblName = rqst.getTableName().toLowerCase();
+    Long validWriteId;
+    Long largestWriteId, smallestWriteId;
+    List<TxnToWriteId> txnToWriteIds = rqst.getTxnToWriteIdList();
+    assert (txnToWriteIds != null);
+
+    // Get the next write id for the table and check if its in consistent with the input txntowriteids.
+    String s = sqlGenerator.addForUpdateClause(
+            "select nwi_next from NEXT_WRITE_ID where nwi_database = " + quoteString(dbName)
+                    + " and nwi_table = " + quoteString(tblName));
+    LOG.debug("Going to execute query <" + s + ">");
+    rs = stmt.executeQuery(s);
+    if (rs.next()) {
+      validWriteId = rs.getLong(1);
+    } else {
+      // All input write ids should be greater than this
+      validWriteId = 0L;
+    }
+
+    int size = txnToWriteIds.size();
+    for (TxnToWriteId txnToWriteId :  txnToWriteIds) {
```

Comment:
It is safe to assume the txnids and writeids in txnToWriteIds are in sorted (asc) order and in sequence. As streaming ingest batch allocation follow this rule. Mention this assumption in the method comment for the metastore thrift API. Also, in the request structure. This assumption will avoid unnecessary traversals.